bootstrap-joining-demo/aws-api/ec2-tag-based/src/it/scala/akka/cluster/bootstrap/IntegrationTest.scala

```diff
@@ -0,0 +1,233 @@
+        new Parameter().withParameterKey("InstanceCount").withParameterValue(instanceCount.toString),
+        new Parameter().withParameterKey("InstanceType").withParameterValue("m3.xlarge"),
+        new Parameter().withParameterKey("KeyPair").withParameterValue("none"),
+        new Parameter().withParameterKey("Purpose").withParameterValue(s"demo-$buildId")
+      )
+
+    awsCfClient.createStack(createStackRequest)
+
+    val describeStacksRequest = new DescribeStacksRequest().withStackName(stackName)
+
+    var dsr: DescribeStacksResult = null
+
+    def conditions: Boolean = (dsr.getStacks.size() == 1) && {
+      val stack = dsr.getStacks.get(0)
+      stack.getStackStatus == StackStatus.CREATE_COMPLETE.toString &&
+      stack.getOutputs.size() >= 1 &&
+      stack.getOutputs.asScala.exists(_.getOutputKey == "AutoScalingGroupName")
+    }
+
+    implicit val patienceConfig: PatienceConfig = createStackPatience
+
+    eventually {
+
+      log.info("CloudFormation stack name is {}, waiting for a CREATE_COMPLETE", stackName)
+
+      dsr = awsCfClient.describeStacks(describeStacksRequest)
+
+      assert(conditions)
+
+    }
+
+    if (conditions) {
+
+      log.info("got CREATE_COMPLETE, trying to obtain IPs of EC2 instances launched")
+
+      val asgName =
+        dsr.getStacks.get(0).getOutputs.asScala.find(_.getOutputKey == "AutoScalingGroupName").get.getOutputValue
+
+      val ips: List[(String, String)] = awsEc2Client
+        .describeInstances(new DescribeInstancesRequest()
+          .withFilters(new Filter("tag:aws:autoscaling:groupName", List(asgName).asJava)))
+          .getReservations
+          .asScala
+          .flatMap((r: Reservation) =>
+              r.getInstances.asScala.map(instance => (instance.getPublicIpAddress, instance.getPrivateIpAddress)))
+          .toList
+          .filter(ips =>
+            ips._1 != null && ips._2 != null) // TODO: investigate whether there are edge cases that may makes this necessary
+
+      clusterPublicIps = ips.map(_._1)
+      clusterPrivateIps = ips.map(_._2)
+
+      log.info("EC2 instances launched have the following public IPs {}", clusterPublicIps.mkString(", "))
+
+    }
+
+  }
+
+  private def readTemplateFromResourceFolder(path: String): String = scala.io.Source.fromResource(path).mkString
+
+  // we need this in order to tell AWS to allow the machine running the integration test to connect to the EC2 instances'
+  // port 19999
+  private def getMyIp: String = {
+    val myIp: Future[(Int, String)] = httpGetRequest("http://checkip.amazonaws.com")
+    val result = Await.result(myIp, atMost = 3 seconds)
+    assert(result._1 == 200, "http://checkip.amazonaws.com did not return 200 OK")
+    result._2
+  }
+
+  test("Integration Test for EC2 Tag Based Discovery") {
+
+    implicit val patienceConfig: PatienceConfig = clusterBootstrapPatience
+    val httpCallTimeout = Timeout(Span(3, Seconds))
+
+    assert(clusterPublicIps.size == instanceCount)
```

Comment:
perhaps use scalatest matchers, which will give better error messages
we often use
```
clusterPublicIps.size should === (instanceCount)
```
but there are many variations

Reply:
@patriknw you are right ... ScalaTest matchers is much better ;-) Will make this change.